proposals/reward-proposal-v2-reward-distribution.md

```diff
@@ -0,0 +1,106 @@
+`K` rewards are distributed from independent `EpochRewardReserve` accounts. The
+leader will inject `K` `DistributeReward` instructions as transactions into the
+block and propagate those transaction down to the network. All nodes on the
+network will execute those transactions and distribute the rewards. In this way,
+all reward distribution will be recoded as transactions in the first `M` blocks
+after reward computation.
+
+### Challenges and Difficulties
+
+The following are a few questions and challenges of the `V2` reward
+distribution approach for further discussion.
+
+1. What are the incentives for the leader to pack reward redeem transactions?
+   What kind of block fees will the leader receive for packing such
+   transactions? Who will pay for the fees for such transactions? Who will be
+   the signer for those transactions? What prevents a malicious leader from not
+   packing those reward distribution transactions?
+
+   There are two ideas to incentivize the leader to include reward redeeming
+   transactions.
+
+   The first idea is to have the network reject the block that doesn't have all
+   the rewards redeemed. This would requires that all the nodes to keep track of
+   what rewards should be redeemed in a block, and verify that the actual block
+   have the expected rewards. Otherwise, the block will be rejected. The
+   downside of this idea is that it will add extra complexity in block replay
+   stage.
+
+   The second idea is to change up the inflation economics to add a "redeemer
+   reward". The validator can claim a portion of it from the merkle tree for
+   properly redeeming a set of stake rewards. For example, if there's 100 SOL in
+   rewards total, we add 1% of that for redeemer rewards, so 1 SOL. Each
+   redemption gives some proportion of the amount redeemed. If a validator
+   redeems 1000 lamports in a block, that's 10 lamports they can claim too. That
+   can even be done at the level of the stake program. This creates a race to
+   reward the biggest whales, but maybe that's ok?
+
+   Also, a quick math shows that for a stake to get 5000 lamports in rewards for
+   one epoch, it needs to be around 0.016 SOL (5000 lamports * 162.5 epochs per
+   year / 5% inflation / 1_000_000_000 lamports per SOL). So if we also restrict
+   the minimum reward earning amount to 1 SOL, and that earns 312,500 lamports /
+   epoch in rewards, with a minimum of 93k lamports at 1.5% inflation. 1% of
+   that is ~1k lamports, so not too far from a transaction fee. If they can pack
+   in 25 accounts in one transaction, hen they still come out on top.
+
+1. What's the performance overhead for the new reward distribution by
+   `DistributeReward` transactions? Is the overhead okay for handling large
+   number of epoch rewards, i.e. 570K?
+
+1. What's the network traffic overhead to inject the reward transactions and
+   broadcast them to all the validator peers?
```

Comment:
Shouldn't this proposal describe the reward transactions which gets executed?
In terms of like a pseudo-code design of the actual structs/enums and what fields they have.

Reply:
yeah. updated.